### PR TITLE
KLD regularized SAC

### LINKED ISSUE #646: Use KL divergence instead entropy for SAC

### PR DESCRIPTION
**Pull request by emailweixu**
Also generate a scalar summary for action mean.

### CODE REVIEW

File: alf/algorithms/prior_actor.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        return self.predict_step(exp, state)
+
+
+@gin.configurable
+class UniformPriorActor(Algorithm):
+    def __init__(self,
+                 observation_spec,
+                 action_spec: BoundedTensorSpec,
+                 debug_summaries=False,
+                 name="PriorActor"):
```
Comment (Haichao-Zhang) [inline alf/algorithms/prior_actor.py]: name="UniformPriorActor"

File: alf/algorithms/sac_algorithm.py [hunk 2]
```diff
@@ -145,6 +145,8 @@ def __init__(self,
                  config: TrainerConfig = None,
                  critic_loss_ctor=None,
                  target_entropy=None,
+                 prior_actor_ctor=None,
+                 target_kld_per_dim=3.,
```
Comment (Haichao-Zhang) [inline alf/algorithms/sac_algorithm.py:155]: May be useful to add a note that target_kld_per_dim  will be used to update self._target_entropy value when prior_actor is used.

File: alf/algorithms/sac_algorithm.py [hunk 3]
```diff
@@ -194,6 +200,14 @@ def __init__(self,
                 callable function, then it will be called on the action spec to
                 calculate a target entropy. If ``None``, a default entropy will
                 be calculated.
+            prior_actor_ctor (Callable): If provided, it will be called using
+                ``prior_actor_ctor(observation_spec, action_spec, debug_summaries=debug_summaries)``
+                to constructor a prior actor. The output of the prior actor is
+                the distribution of the next action. Two prior actors are implemented:
+                ``alf.algorithms.prior_actor.PriorActor`` and
+                ``alf.algorithms.prior_actor.UniformPriorActor``.
+            target_kld_per_dim (float): ``alpha`` is dynamically ajusted so that
```
Comment (Haichao-Zhang) [inline alf/algorithms/sac_algorithm.py]: ajusted -> adjusted

File: alf/algorithms/prior_actor.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+from alf.algorithms.algorithm import Algorithm
+from alf.data_structures import AlgStep, Experience, TimeStep, StepType
+from alf.networks import Network
+from alf.tensor_specs import TensorSpec, BoundedTensorSpec
+
+
+def normcdf(a, b):
+    # 0.7071067811865476 = math.sqrt(0.5)
+    return 0.5 * (
+        torch.erf(0.7071067811865476 * b) - torch.erf(0.7071067811865476 * a))
```
Comment (hnyu) [inline alf/algorithms/prior_actor.py]: use the variable to hold this constant as it appears twice?

File: alf/algorithms/sac_algorithm.py [hunk 5]
```diff
@@ -91,6 +91,12 @@ class SacAlgorithm(OffPolicyAlgorithm):
 
         Haarnoja et al "Soft Actor-Critic Algorithms and Applications", arXiv:1812.05905v2
 
+    In addition to the entropy regularization described in the SAC paper, we
+    also support KL-Divergence regularization if a prior actor is provided.
+    In this case, the training objective is:
+        :math:`E_\pi(\sum_t \gamma^t(r_t + \alpha D_{\rm KL}(\pi(\cdot)|s_t)||\pi^0(\cdot)|s_t)))`
```
Comment (Haichao-Zhang) [inline alf/algorithms/sac_algorithm.py]: ```r_t + \alpha D_{\rm KL}``` -> should be ```r_t - \alpha D_{\rm KL}```

File: alf/algorithms/prior_actor.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+    def rsample(self):
+        raise NotImplementedError()
+
+
+class MixtureSameFamily(td.Distribution):
+    """
+    ``MixtureSameFamily`` was introduced in pytorch 1.5. Here we only implement
+    a subset of its functions.
+    """
```
Comment (hnyu) [inline alf/algorithms/prior_actor.py:86]: Description of what this class does

**Review by Haichao-Zhang — COMMENTED** [review]
LGTM. Just two more minor comments on the newly added commit.

File: alf/algorithms/prior_actor.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+    def log_prob(self, x):
+        """Log-probability of sample ``x``."""
+        x = x.unsqueeze(-1 - self._event_ndims)
+        log_prob_x = self._component_distribution.log_prob(x)  # [S, B, k]
+        log_mix_prob = self._mixture_distribution.logits  # [B, k]
+        return torch.logsumexp(log_prob_x + log_mix_prob, dim=-1)  # [S, B]
+
+
+@gin.configurable
+class PriorActor(Algorithm):
```
Comment (hnyu) [inline alf/algorithms/prior_actor.py]: The name of this class could be better. Right now it sounds like a general class of "uniformprioractor", but it's actually quite different.

File: alf/algorithms/sac_algorithm.py [hunk 8]
```diff
@@ -91,6 +91,12 @@ class SacAlgorithm(OffPolicyAlgorithm):
...
+    where :math:`pi^0` is the prior actor.
```
Comment (hnyu) [inline alf/algorithms/sac_algorithm.py]: This paragraph should be moved to below "points different"

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/algorithms/prior_actor.py (hunk 4 above)
Comment (emailweixu) [reply alf/algorithms/prior_actor.py]: Changed. Though I believe it will be a little bit slower.

File: alf/algorithms/prior_actor.py (hunk 6 above)
Comment (emailweixu) [reply alf/algorithms/prior_actor.py:86]: Updated

File: alf/algorithms/prior_actor.py (hunk 7 above)
Comment (emailweixu) [reply alf/algorithms/prior_actor.py]: updated

File: alf/algorithms/sac_algorithm.py (hunk 8 above)
Comment (emailweixu) [reply alf/algorithms/sac_algorithm.py]: moved

**Review by Haichao-Zhang — DISMISSED** [review]

**Review by hnyu — APPROVED** [review]

**Pull request merged**